### ISSUE TITLE
Syncing is slow in 0.16.0rc3

### ISSUE THREAD

**Issue by clayop** [steemit/steem#614]
One of my node is syncing very slowly. It is doing for a couple of days (I think it needs several hours more). Meanwhile, my other node that is dedicated to steemd works good.

The slow node specs are: 4-core VPS with 16G memory and 60G SSD
Running MongoDB and some python3 scripts

Here's result from top
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                          
 1989 clayop    20   0 33.484g 0.010t 0.010t R 122.3 42.0 668:49.66 steemd                                           
 2478 clayop    20   0   90716  27156   9812 S   0.7  0.1   7:14.91 python3                                          
 1330 mongodb   20   0 6405016 5.940g  21928 S   0.3 23.3   7:29.51 mongod                                           
```

and free -m
```
              total        used        free      shared  buff/cache   available
Mem:          26127        6352         733           8       19041       19371
Swap:             0           0           0
```

**Comment by abitmore** [comment]
Check your IO stats (the `%Cpu ... wa` in `top`)? I have been testing with an AWS vps with "general purpose" SSD (caution: with some IO limitations), so the SSD was almost "killed" in the first hour, and in following hours I experienced terrible IO issue. I'll use ram-disks for full sync and perhaps replay when needed.

**Comment by mvandeberg** [comment]
Are you using the default flush interval?

**Comment by mvandeberg** [comment]
@abitmore How much memory did your AWS instance have?

**Comment by clayop** [comment]
I'm using default.

**Comment by theoreticalbts** [comment]
Please post output of this command:

    grep ^ /proc/sys/vm/dirty*

If you have not done so, try changing these kernel parameters [as suggested in the release announcement](https://steemit.com/steem/@steemitblog/steem-0-16-0-official-release).

**Comment by clayop** [comment]
```
/proc/sys/vm/dirty_background_bytes:0
/proc/sys/vm/dirty_background_ratio:10
/proc/sys/vm/dirty_bytes:0
/proc/sys/vm/dirty_expire_centisecs:3000
/proc/sys/vm/dirty_ratio:20
/proc/sys/vm/dirtytime_expire_seconds:43200
/proc/sys/vm/dirty_writeback_centisecs:500
```

**Comment by mvandeberg** [comment]
https://steemit.com/steem/@steemitblog/steem-0-16-0-official-release

**Comment by abitmore** [comment]
I was aware of the kernel parameter tweaks but just a little bit hesitate to do so, so didn't apply them while testing. I was using default flush intervals (100K).

```
/proc/sys/vm/dirty_background_bytes:0
/proc/sys/vm/dirty_background_ratio:10
/proc/sys/vm/dirty_bytes:0
/proc/sys/vm/dirty_expire_centisecs:3000
/proc/sys/vm/dirty_ratio:20
/proc/sys/vm/dirty_writeback_centisecs:500
```

The AWS instance I was testing with has 15G RAM and 50G "General Purpose" SSD (which means it can burst to 3000 IOPS for 30 minutes, after that, it will fall to 150 IOPS thus caused bad performance)

**Comment by theoreticalbts** [comment]
Looks like you did not tweak all the kernel parameters, in particular `dirty_ratio` / `dirty_background_ratio` appear to be the default values for both @abitmore and @clayop .

From [this blog post](https://lonesysadmin.net/2013/12/22/better-linux-disk-caching-performance-vm-dirty_ratio/):

"vm.dirty_ratio is the absolute maximum amount of system memory that can be filled with dirty pages before everything must get committed to disk. When the system gets to this point all new I/O blocks until dirty pages have been written to disk."

Basically dirty_ratio times system RAM is the max memory you can actually use for the "hot" part of the shared file that's actively being written, in your case it's 3.2GB.  I've noticed while reindexing on my own system that about 6.5GB remains dirty after even intense I/O bursts.  This tells me that 6.5GB is "hot" based on the actual blockchain activity.  If you have 6.5 GB of "hot" data but only 3.2 GB RAM to hold it, you'll get a ton of churn.

So for right now the rule of thumb would be, anytime your dirty_ratio times system RAM is less than ~8GB you will probably have a bad time.  Therefore tweaking these kernel parameters is essential for Linux users.

**Comment by clayop** [comment]
I am testing how the tweaks make different replaying time (it was so slow too). So far seems good.

**Comment by clayop** [comment]
I changed parameters and turned off mongodb that was occupying 40% of my memory. The I/O issue was quite mitigated (I first changed three parameters except `dirty_expire_centisecs` but it didn't work well. After I changed `dirty_expire_centisecs` to 6000 it works).

I also found the difference in syncing speed is different IOPS capability depending on VPS disk size. My main node with 100G SSD has 3000 random IOPS limit 48MB/s throughput limit, while a sub-node has 1500/24. But I feel syncing is much faster than before I changed the parameters.
